pyogrio/_vsi.pyx

```diff
@@ -118,23 +125,152 @@ cdef read_vsimem_to_buffer(str path, object out_buffer):
+    - a toplevel file (directly in /vsimem/).
+    - a file in a directory, with possibly some sibling files.
+    - a zip file, which apparently is reported as a directory by VSI_ISDIR.
+
+    Except for the first case, the toplevel directory (direct subdirectory of
+    /vsimem/) will be determined and will be removed recursively.
 
     Additional VSI handlers may be chained to the left of /vsimem/ in path and
     will be ignored.
 
+    Even though it is only meant for "internal use", the function is declared
+    as cpdef, so it can be called from tests as well.
+
     Parameters:
     -----------
     path : str
         path to in-memory file
+
     """
+    cdef VSIStatBufL st_buf
 
     if "/vsimem/" not in path:
-        return
+        raise ValueError(f"Path is not a /vsimem/ path: '{path}'")
+
+    # Determine the toplevel directory of the file
+    mempath_parts = path.split("/vsimem/")[1].split("/")
+    if len(mempath_parts) == 0:
+        raise OSError("Removing /vsimem/ is not supported")
+
+    toplevel_path = f"/vsimem/{mempath_parts[0]}"
+
+    if not VSIStatL(toplevel_path.encode("UTF-8"), &st_buf) == 0:
+        raise FileNotFoundError(f"Path does not exist: '{path}'")
+    if VSI_ISDIR(st_buf.st_mode):
+        errcode = VSIRmdirRecursive(toplevel_path.encode("UTF-8"))
+    else:
+        errcode = VSIUnlink(toplevel_path.encode("UTF-8"))
+
+    if errcode != 0:
+        raise OSError(f"Error removing '{path}': {errcode=}")
+
+
+def ogr_vsi_listtree(str path, str pattern):
+    """Recursively list the contents in a vsi directory.
+
+    An fnmatch pattern can be specified to filter the directories/files
+    returned.
+
+    Parameters:
+    -----------
+    path : str
+        Path to the vsi directory to be listed.
+    pattern : str
+        Pattern to filter results, in fnmatch format.
 
-    root = "/vsimem/" + path.split("/vsimem/")[1].split("/")[0]
-    VSIRmdirRecursive(root.encode("UTF-8"))
+    """
+    cdef const char *path_c
+    cdef int n
+    cdef char** papszFiles
+    cdef VSIStatBufL st_buf
+
+    try:
+        path_b = path.encode("UTF-8")
+    except UnicodeDecodeError:
+        path_b = path
+    path_c = path_b
+    if not VSIStatL(path_c, &st_buf) == 0:
+        raise FileNotFoundError(f"Path does not exist: '{path}'")
+    if not VSI_ISDIR(st_buf.st_mode):
+        raise NotADirectoryError(f"Path is not a directory: '{path}'")
+
+    papszFiles = VSIReadDirRecursive(path_c)
+    n = CSLCount(<CSLConstList>papszFiles)
+    files = []
+    for i in range(n):
+        files.append(papszFiles[i].decode("UTF-8"))
+    CSLDestroy(papszFiles)
+
+    # Apply filter pattern
+    if pattern is not None:
+        files = fnmatch.filter(files, pattern)
+
+    # Prepend files with the base path
+    if not path.endswith("/"):
+        path = f"{path}/"
+    files = [f"{path}{file}" for file in files]
+    
+    return files
+
+
+def ogr_vsi_rmtree(str path):
+    """Recursively remove vsi directory.
```

Comment:
```suggestion
    """Recursively remove VSI in-memory directory.
```

Reply:
Should these new `ogr_vsi_*` functions be limited to `/vsimem/` paths, or are you intending that they should work for some of the other VSI interfaces as well?

Reply:
I gave this some thought, and couldn't really find a good reason to limit it to vsimem support. Storing files in the cloud is also popular I suppose, so I can imagine it might be useful to just expose the full GDAL capabilities?

Reply:
Further discussed in another comment...